### DISCUSSION TITLE
Can brew be used with git bisect?

### DISCUSSION THREAD

**Discussion by jmckitrick** [Homebrew/discussions#4649] [Writing Formulae/Casks]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.28
ORIGIN: https://github.com/Homebrew/brew
HEAD: 2b4cfb033757b63ab0b128d01c58db76480956f9
Last commit: 8 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 1d8c58a0aad7953501bd69df9f5014a6dfbf199e
Core tap last commit: 5 months ago
Core tap branch: master
Core tap JSON: 14 Jul 10:38 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_INSTALL_BADGE: ☕️
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.41.0 => /opt/homebrew/bin/git
Curl: 7.88.1 => /usr/bin/curl
macOS: 13.4.1-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: Calling plist_options is deprecated! Use service.require_root instead.
Please report this issue to the elastic/tap tap (not Homebrew/brew or Homebrew/homebrew-core), or even better, submit a PR to fix it:
  /opt/homebrew/Library/Taps/elastic/homebrew-tap/Formula/elasticsearch-full.rb:68

Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have an unnecessary local Cask tap.
This can cause problems installing up-to-date casks.
Please remove it by running:
  brew untap homebrew/cask

Warning: You have an unnecessary local Core tap!
This can cause problems installing up-to-date formulae.
Please remove it by running:
 brew untap homebrew/core
```

### Description of issue

I found a serious issue in emacs 30 that does not exist in emacs 29. I want to use git bisect to narrow down when this problem appeared. I want to build emacs 30 at a specific git commit, run a test, and continue to bisect.

I'm being told brew does not support this. Is that true? What are my other options?

**Comment by gromgit** [comment]
`git bisect` works exclusively with local source trees, and requires the user to run the post-bisect builds themselves. Homebrew doesn't build from local source trees at all, so if you're thinking of iterating `git bisect` / `brew install -s` / `brew test`, that's not gonna work. At all.

Your best option is to write your own config/build/test script.

**Reply by osalbahr** [reply]
> Homebrew doesn't build from local source trees at all

Well, yes and no. It most certainly does if you set [`HOMEBREW_NO_INSTALL_FROM_API`](https://docs.brew.sh/FAQ#can-i-edit-formulae-myself).

That being said, I think [creating a tap](https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap) would be a much better route for needing to use an old version.

**Reply by gromgit** [reply]
> > Homebrew doesn't build from local source trees at all
> 
> Well, yes and no. It most certainly does if you set [`HOMEBREW_NO_INSTALL_FROM_API`](https://docs.brew.sh/FAQ#can-i-edit-formulae-myself).

If `HOMEBREW_NO_INSTALL_FROM_API` is involved, that means you're `git bisect`ing _Homebrew formulae_, not the Emacs source tree. For that to work for the OP, you'd need a tap with an Emacs formula commit _that checks out a corresponding Emacs source tree commit of interest_, otherwise the OP's requirement:

> I want to build emacs 30 *at a specific git commit*

simply can't be done with `brew`. ("OK, I've got a formula for Emacs 30, and one for Emacs 29, now where's the one for Emacs commit `fd23aeb`? Why isn't there one?!?!")

Also bear in mind that Homebrew would be re-copying the entire Emacs build tree into a new sandbox at every `git bisect` stage. You're basically re-running the entire Homebrew build process from scratch each time, not simply recompiling Emacs sources that are already checked out at the appropriate commit. (I assume the original cloned source tree is retained in Homebrew cache; if not, this just went from "painfully slow" to "please kill me now".)

**Reply by osalbahr** [reply]
Oh, yeah you’re right. Sorry I misunderstood

**Comment by jmckitrick** [comment]
I think I'll try my hand at making my own tap, and that should do the trick.

**Reply by osalbahr** [reply]
Good luck! I hope that works out
